app/src/main/java/org/wikipedia/editactionfeed/provider/MissingDescriptionProvider.kt

```diff
@@ -13,27 +13,53 @@ import org.wikipedia.dataclient.restbase.page.RbPageSummary
 import org.wikipedia.page.PageTitle
 import org.wikipedia.wikidata.Entities
 import java.util.*
+import kotlin.collections.ArrayList
 
 object MissingDescriptionProvider {
+    val articlesWithMissingDescriptionCache : Stack<String> = Stack()
+    var articlesWithMissingDescriptionCacheLang : String = ""
+    val articlesWithTranslatableDescriptionCache : Stack<Pair<PageTitle, PageTitle>> = Stack()
+    var articlesWithTranslatableDescriptionCacheFromLang : String = ""
+    var articlesWithTranslatableDescriptionCacheToLang : String = ""
+
     // TODO: add a maximum-retry limit -- it's currently infinite, or until disposed.
 
     @Deprecated("Remove when the new API is deployed to production.")
     fun getNextArticleWithMissingDescription(wiki: WikiSite): Observable<RbPageSummary> {
-        return ServiceFactory.get(wiki).randomWithPageProps
-                .map<List<MwQueryPage>> { response ->
-                    val pages = ArrayList<MwQueryPage>()
-                    for (page in response.query()!!.pages()!!) {
-                        if (page.pageProps() == null || page.pageProps()!!.isDisambiguation || !TextUtils.isEmpty(page.description())) {
-                            continue
+        var cachedTitle = ""
+        synchronized(articlesWithMissingDescriptionCache) {
+            if (articlesWithMissingDescriptionCacheLang != wiki.languageCode()) {
+                // evict the cache if the language has changed.
+                articlesWithMissingDescriptionCache.clear()
+            }
+            if (articlesWithMissingDescriptionCache.size > 0) {
+                cachedTitle = articlesWithMissingDescriptionCache.pop()
+            }
+        }
+
+        return (if (!TextUtils.isEmpty(cachedTitle)) Observable.just(cachedTitle) else
```

Comment:
I am a little bit confusing that the should `cachedTitle`be a type of `RbPageSummary`?

Reply:
Nope, we want to cache the *seed* titles from which the RbPageSummary will be retrieved. Remember that the initial call gets up to fifty (50) random articles, but only uses one of them for getting the `RbPageSummary` (it throws away the other 49).  This optimization keeps the other 49, and reuses them in subsequent calls.

Reply:
I see. I thought it is a `return` value of the method and didn't notice that is actually a part of the `return` before calling `flatMap`